File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */


class Solution {
public:
    int INF = 1e9;
    vector<int> nodesBetweenCriticalPoints(ListNode* head) {

        ListNode* temp = head;
        vector<int> arr;
        while (temp != nullptr) {
            arr.push_back(temp->val);
            temp = temp->next;
        }

        if (arr.size() <= 2) {
            return {-1, -1};
        }
        
        vector<int> ans = {INF, 0};
        int firstCritical = 0;
        int prevCritical = 0;
        for (int i = 1; i < arr.size() - 1; i++) {
            if ((arr[i - 1] < arr[i] && arr[i] > arr[i + 1]) ||
                (arr[i - 1] > arr[i] && arr[i] < arr[i + 1])) {
                if (firstCritical == 0) {
                    firstCritical = i;
                    prevCritical = i;
                } else {
                    ans[0] = min(ans[0], i - prevCritical);
                    ans[1] = max(ans[1], i - firstCritical);

                    prevCritical = i;
                }
            }
        }

        if (ans[1] == 0) {
            return {-1, -1};
        }

        return ans;
    }
};

File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    int INF = (int) 1e9;

    public int[] nodesBetweenCriticalPoints(ListNode head) {
        int[] ans = {INF, 0};
        
        int dist = 1;
        int firstCritical = 0;
        int prevCritical = 0;

        for (ListNode curr = head.next, prev = head; curr.next != null; curr = curr.next, prev = prev.next, dist += 1) {
            if ((prev.val < curr.val && curr.val > curr.next.val) || (prev.val > curr.val && curr.val < curr.next.val)) {
                if (firstCritical == 0) {
                    firstCritical = dist;
                    prevCritical = dist;
                } else {
                    ans[0] = Math.min(ans[0], dist - prevCritical);
                    ans[1] = Math.max(ans[1], dist - firstCritical);

                    prevCritical = dist;
                }
            }
        }

        if (ans[1] == 0) {
            return new int[]{-1, -1};
        }

        return ans;
    }
}